### DISCUSSION TITLE
process.env not defined with test files.

### DISCUSSION THREAD

**Discussion by lefuncq** [cypress-io/cypress#26200] [Questions and Help]
Hi,

I have an e2e test file called `ConnectWallet.cy.js` containing the following code:
```
import { WALLETS } from '@/constants';

const elements = {
  ...Object.values(WALLETS).map((wallet) => ({
    [`walletButton${wallet}`]: `[data-testid="connect-wallet-container-wallet-button-${wallet}"]`,
  })),
};

describe('ConnectWallet', () => {
  it('Visitor', () => {
    cy.visit('http://localhost:3000');
  });
});
```
`constants` is a file containing constants shared amongst various pages and components of my NextJS App, and some of the values there, are defined using `process.env`, thus raising a "Referrence Error: process is not defined" error within Cypress's test browser (_The following error originated from your test code, not from Cypress._).
I've been searching everywhere but couldn't find a solution that works:
- Setting the env within `cypress.config.js` will just expose my env variables to `Cypress.env` which I can't use within `constants.js` since `Cypress.env` is, legitimately, undefined when running my next server.
- I've tried loading process.env using `on('file:preprocessor', ...` but couldn't find how to set up the web pack so that it would "resolve/load" my process.env inside of test files.
Isn't there any workaround to that issue?

**Comment by portyu9** [comment]
Hi @lefuncq 

The problem is that the shared module is relying directly on a **Next.js build-time feature**.

Your Cypress spec runs in the browser. When it imports `@/constants`, that module becomes part of the Cypress test bundle. A normal browser does not have Node's global:

```js
process
```

Next.js normally handles references such as:

```js
process.env.NEXT_PUBLIC_RPC_URL
```

by replacing them while building the application. That transformation is not something a generic shared module can assume every consumer will perform.

I would make the shared module independent of both Next.js and Cypress:

```js
// wallets.js
export function createWallets({ rpcUrl }) {
  return {
    // use rpcUrl here
  }
}
```

Then the Next.js application supplies its value:

```js
const wallets = createWallets({
  rpcUrl: process.env.NEXT_PUBLIC_RPC_URL,
})
```

and Cypress supplies its own value.

With current Cypress, a public/non-sensitive configuration value can be exposed and read synchronously:

```js
const wallets = createWallets({
  rpcUrl: Cypress.expose('RPC_URL'),
})
```

For a **secret**, do not put it into a browser-side shared constants module. Current Cypress uses `cy.env()` for sensitive values, or you can keep secret-dependent work in Node via `cy.task()`.

So the important design change is: **do not have a supposedly shared module directly depend on `process.env`**. Inject the environment-specific values from the Next.js or Cypress entry point instead.

Please mark this answer as accepted if it helped, thank you.
